runhouse/servers/env_servlet.py

```diff
@@ -227,61 +261,66 @@ def _get_env_cpu_usage(self, cluster_config: dict = None):
-                        "nvidia-smi",
-                        "--query-compute-apps=pid,gpu_uuid,used_memory",
-                        "--format=csv,nounits",
-                    ],
-                    stdout=subprocess.PIPE,
-                )
-                .stdout.decode("utf-8")
-                .strip()
-                .split("\n")
-            )
-            for i in range(1, len(env_gpu_usage)):
-                single_env_gpu_info = env_gpu_usage[i].strip().split(", ")
-                if int(single_env_gpu_info[0]) == env_servlet_pid:
-                    env_used_memory = env_used_memory + int(single_env_gpu_info[-1]) * (
-                        1024**2
-                    )
-            if env_used_memory > 0:
-                env_gpu_usage = {
-                    "used_memory": env_used_memory,  # in bytes
-                    "total_memory": total_gpu_memory,  # in bytes
-                }
-            else:
-                env_gpu_usage = {}
-        except subprocess.CalledProcessError as e:
-            logger.error(f"Failed to get GPU usage for {self.env_name}: {e}")
-            env_gpu_usage = {}
+        if collected_gpus_info is None or not collected_gpus_info[0]:
+            return None
+
+        return get_gpu_usage(
+            collected_gpus_info=collected_gpus_info, servlet_type=ServletType.env
+        )
 
-        return env_gpu_usage
+    def _collect_env_gpu_usage(self):
+        """periodically collects env gpu usage"""
+
+        pynvml.nvmlInit()  # init nvidia ml info collection
+
+        while True:
+            try:
+                gpu_count = pynvml.nvmlDeviceGetCount()
+                with self.lock:
+                    if not self.gpu_info:
+                        self.gpu_info = {device: [] for device in range(gpu_count)}
+
+                    for gpu_index in range(gpu_count):
+                        handle = pynvml.nvmlDeviceGetHandleByIndex(gpu_index)
+                        processes = pynvml.nvmlDeviceGetComputeRunningProcesses_v3(
+                            handle
+                        )
+                        memory_info = pynvml.nvmlDeviceGetMemoryInfo(handle)
+                        if processes:
+                            for p in processes:
+                                if p.pid == self.pid:
+                                    used_memory = p.usedGpuMemory  # in bytes
+                                    total_memory = memory_info.total  # in bytes
+
+                                    updated_gpu_info = self.gpu_info[gpu_index]
+                                    # to reduce cluster memory usage (we are saving the gpu_usage info on the cluster),
+                                    # we save only the most updated gpu usage. If for some reason the size of updated_gpu_info is
+                                    # too big, we remove the older gpu usage info.
+                                    if len(updated_gpu_info) + 1 > MAX_GPU_INFO_LEN:
```

Comment:
a little confused about this- the gpu info will store data from the last x minutes, but we only show the most recent? why would this gpu info hold older data in the first palce?